File: GP_Cpp/GPNode.cpp
/*
 * GPNode.cpp
 *
 *  Created on: 8 feb 2021
 *      Author: antonio
 */

#include "GPNode.h"

GPNode::GPNode(string info, int arity, NodeType type) : GPNode(info, arity)
{
   this->type =type;
}

GPNode::GPNode(string info, int arity) {
   this->info = info;
   this->arity = arity;
   this->type = (arity == 0)? TERMINAL : FUNCTION;
   children.resize(this->arity, nullptr);
   value = 0.0;
}

GPNode::~GPNode() {}

int GPNode::getArity() {
   return children.size();
}

void GPNode::setChild(int i, GPNode* child) {
   children[i] = child;
}

GPNode* GPNode::getChild(int i) {
   if (arity == 0)
      return nullptr;
   else
      return children[i];
}

string GPNode::getInfo() {
   return info;
}

bool GPNode::isTerminal() {
   return (arity == 0);
}

bool GPNode::isFunction() {
   return (arity > 0);
}

File: GP_Python/gpNode.py
from __future__ import annotations
from enum import Enum

#Node Type
class NType(Enum):
    FUNCTION    = 1
    TERMINAL    = 2
    LOGIC       = 3
    VARIABLE    = 4
    CONSTANT    = 5
    ACTION      = 6

class Node:
    def __init__(self, info, arity, type=None):
        self.info = info
        self.arity = arity
        if type != None: 
            self.type = type
        else: 
            self.type = NType.TERMINAL if arity == 0 else NType.FUNCTION
        
        self.children = [None]*arity
        self.value = 0.0

    def getArity(self) -> int:
        return len(self.children)

    def setChild(self, i, child):
        self.children[i]=child

    def getChild(self, i) -> Node:
        if self.arity == 0:
            return None
        else:
            return self.children[i]

    def isTerminal(self)->bool:
        return (self.arity == 0)

    def isFunction(self)->bool:
        return (self.arity > 0) 

